### DISCUSSION TITLE
add resources via shortcode

### DISCUSSION THREAD

**Discussion by jmbuhr** [quarto-dev/quarto-cli#1938] [Q&A] [answered]
We have a shortcode that adds code which loads user-supplied external files. In order for this to work for a site, the external files have to be added to the resources (https://quarto.org/docs/websites/website-tools.html#site-resources).

Is it possible to do this from within the lua code of the shortcode?

e.g. in `{{< mol-url ./www/7sgl.pdb >}}` at https://jmbuhr.de/quarto-molstar/#local-pdb-file we have to add `./www/7sgl.pdb` to the resources so that it will be copied.

refs: https://github.com/jmbuhr/quarto-molstar/issues/7

**Comment by jjallaire** [comment, marked as answer]
Yes definitely! You want to use HTML dependencies and/or format resources as described here: https://quarto.org/docs/extensions/lua.html#dependencies

**Reply by jmbuhr** [reply]
Does this also work for user-supplied path's at runtime, relative to the quarto document? I assumed `quarto.doc.addHtmlDependency` only worked for files brought by the extension, as it talked about the resource path being "relative to Lua script"

**Comment by jjallaire** [comment]
Your extension code runs in the same working directory as the input document,

If you have trouble making a given scenario work LMK and we'll make sure there is a hook to make it so.

**Reply by jmbuhr** [reply]
Maybe I am missing something obvious. I just tried using it to copy the filepath supplied to the shortcode, but I get a `No such file or directory (os error 2)`.

This is the attempt: https://github.com/jmbuhr/quarto-molstar/pull/9

specifically these lines: https://github.com/jmbuhr/quarto-molstar/blob/1cbdb173809905bab05dd55034c81211615e8a03/_extensions/molstar/molstar.lua#L56-L68

and the path is e.g. something like `./www/7sgl.pdb` (which does exist in the repo).
So I think this only works for resources supplied by the extension.
What I am essential looking for is a lua binding for https://quarto.org/docs/websites/website-tools.html#site-resources

**Comment by jjallaire** [comment]
Okay, I think we need to extend the `addHtmlDependency` API to allow specification of those sorts of resources. @cscheid is working on that now....

**Reply by jmbuhr** [reply]
Hi there, did this make it's way to a branch, yet? I saw a couple of commits in that direction, but not entirely the same.
(Bumping this to keep track, since it is already flagged as answered)

**Reply by cscheid** [reply]
I think you're specifically looking for this: https://github.com/quarto-dev/quarto-cli/commit/00091afdd00b64545887ee14984fdba5bb8f04b7

The Lua function `attachToDependency` will add a new file to an existing dependency. This is a new feature, designed so that different pages can attach different files to the same dependency. The idea is that one page will have (say) `{{< mol-url ./www/7sgl.pdb >}}` and another page will have `{{< mol-url ./www/9sgl.pdb >}}`. If your shortcode declares an HTML dependency called `mol-files`, then you can attach these different resources to the dependency and they will be included.

**Reply by jmbuhr** [reply]
I played around with it a bit, but was unsuccessful in finding a solution that works for both sites/books and standalone documents.

I believe the function can only copy paths to the `site_libs` dir of a site (ideally if would copy to the root directory, which would be the `_site` or `docs` folder for a project or the working directory for a single document (i.e. not move the file at all))

```
local function addResource(path)
  local m = path:match('^https?://')
  -- location of the path relative to this lua script
  local location = pandoc.path.join{'../..' ,path}
  if m == nil then
    quarto.doc.attachToDependency('molstar', {name = location, path = location})
  end
  local newPath = pandoc.path.join{'site_libs', path}
  return newPath
end
```

The directory ups-and-downs feel weird (due to things being relative to the lua file instead of the working directory of the quarto file at runtime) , but this works for a project: https://github.com/jmbuhr/molstar-test
However, it fails for single documents, since they generate no `site_libs` folder and instead of a `<filename>_files` folder (e.g. the example files in the extension repo: https://github.com/jmbuhr/quarto-molstar/tree/copy-resources).

Maybe I can specify a bit more what I think is needed:
Take a hypothetical extension, that includes images into html (maybe it does some fancy filtering on top so it's not the regular markdown for including images, but has no external dependencies):
`{{< my-image ./path/to/image.png >}}`
Because this is not handled by pandocs regular image machinery quarto and pandoc don't know to copy the file to where the html output file ends up. Quarto does know about the file when we add
```
resources:
  - `path/to/image.png`
```
to the file or project, but what if we want to automate exactly this part, the addition of resources to the output folder? Ideally in such a way that a file relative to the source document ends up having the same relative when viewed from the output document.

**Comment by jjallaire** [comment]
@dragonstyle Let's discuss this realtime later on this morning
